Fixes a problem with posts losing their local changes.

diegoreymendez opened the pull request:
Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/4904 and https://github.com/wordpress-mobile/WordPress-iOS/issues/404 (only the first part of it).

Also makes local revisions much more resilient to these kind of issues in the future.  Local changes that are not in the server will also be tagged now as "Local" regardless of if the post was previously published or not.

**Test 1:**
1. Launch the app.
2. Create a new post.
3. Publish it.
4. Edit the post.
5. Disconnect from the internet.
6. Update the post, you should see a failure message.

Expected behavior: the post should be marked as local in the list of posts.  The original should not be visible.

**Test 2:**
1. Follow all the steps from test 1.
2. Once you see the post marked as "Local", pull down the list to refresh.  The post should remain unaffected by the action.
3. Edit the original post using your web browser and save it.
4. Pull down to refresh the list of posts again.

Expected behaviour: the local post should not be affected by either the pull to refresh action, or by changes applied through the web interface.

**Test 3:**
Make sure that posts tagged as local always tell you there are unsaved changes when you open them.

**Known issues:**
There are some refreshing issues in the list of posts that will be handled in a different PR.  The refreshing issues existed before this PR but may seem to come up more often due to the change in GUI logic.

Needs review: @aerych

diegoreymendez commented:
Ping @astralbodies as requested.

kurzee commented:
@diegoreymendez I'd be happy to take a look for your upcoming known issues PR. I recently did a lot of work dealing with the posts refreshing implementation.

aerych commented:
@kurzee The more the merrier!!

aerych commented on WordPress/Classes/Models/Post.m:
```diff
@@ -196,18 +196,19 @@ - (NSString *)contentPreviewForDisplay
-        statusString = nil;
-    }
-    if (![self hasRemote]) {
+    NSString *statusString = nil;
```
Fwiw, the nil assignment is unnecessary and the declared property resolves to nil with out it.

aerych commented on WordPress/Classes/Models/Post.m:
```diff
@@ -196,18 +196,19 @@ - (NSString *)contentPreviewForDisplay
         if (statusString) {
             statusString = [NSString stringWithFormat:@"%@, %@", statusString, localOnly];
         } else {
             statusString = localOnly;
```
The way this is composed, `statusString` will always be nil for a revision thus only the `else` clause will be executed. 
This is problematic for scheduled posts, and posts that are pending review as we would still want to display those statuses.

aerych replied:
Should the status we show make a distinction between posts that are "local only" and posts that have local changes? Just a thought.

diegoreymendez replied:
> The way this is composed, statusString will always be nil for a revision thus only the else clause will be executed. 

The idea behind this PR is that only non-uploaded changes are marked as "revisions".  So whenever something is marked as a revision, it means there are local changes that were never uploaded to the server.

> This is problematic for scheduled posts, and posts that are pending review as we would still want to display those statuses.

I'm testing it in the simulator here and it's working as I'd expect it to.  Some example cases:

1. Post is immediate, but failed to upload: "Local"
2. Post is immediate, and uploaded successfully: nil.
3. Post is scheduled, but failed to upload: marked "Local"
4. Post is scheduled, and uploaded successfully: "Scheduled"
5. Post is awaiting review, but failed to upload: "Local"
6. Post is awaiting review, and uploaded successfully: "Awaiting review"

The idea of the "Local" tag is to let the user know he will potentially loose those changes if he doesnt try to upload them again.

diegoreymendez replied:
> Should the status we show make a distinction between posts that are "local only" and posts that have local changes? Just a thought.

Could be... the main idea behind this PR is that whether all of the content needs to be uploaded or just some of it - the user should know he's risking to lose changes if, for example, they log out.

diegoreymendez replied:
Maybe "Has local changes", or some other label, would be better to describe the situation?

diegoreymendez replied:
Regarding the trashed posts, I'll make sure to correct the label.

aerych replied:
I think my expectation was  that rather than replacing the existing status completely, the "Local" would be an additional status.  For example, this is what currently is displayed when creating a scheduled post while offline.  Notice that both Scheduled, and Local are displayed.

<img width="379" alt="screen shot 2016-03-14 at 4 16 14 pm" src="https://cloud.githubusercontent.com/assets/1435271/13760262/3df40fd8-ea00-11e5-8304-f3d604762efa.png">

The changes in this branch replace the status completely with "local" which seems counter to our existing design. This is what I was referring to when I said it seemed problematic for certain post statuses. 

Maybe in the case of local edits to existing posts, it could continue to display whatever the status was (if its a status we display) and have "Local Changes" appended?

diegoreymendez replied:
Understood.  I'm glad I asked you to check this PR. :)

aerych commented:
> There are some refreshing issues in the list of posts that will be handled in a different PR

Is this perhaps referring to the failure message when publishing while offline?  I was noticing how that was a bad experience.   Maybe related, there was also the prompt about "unsaved changes" when returning to the editor while still offline.

aerych commented:
Scenario - Trashing a post with local edits:
- Open the post list and edit an existing post. 
- Disable your internet connection (airplane mode) and save your edits.
- Notice the post list's preview will show the change if it was made within the first couple of lines.  Notice also the "Local" status. 
- Restore the internet connection and trash the post. 
- Switch to the Trash filter and find the post in the list of trashed posts. 
- Notice that it still shows the Local status but the preview no longer shows the change you made. 
- Restore the post. 
- Return to the published list and find the post. 
- Notice the post still shows the "Local" status. 
- Edit the post.  
- Notice that your local changes are gone but the app still treats the post as having unsaved edits.

diegoreymendez commented:
Hold on on this PR for a bit.  I'll be reviewing the trashing logic and will let you know @aerych when it's ready again.

diegoreymendez commented:
@aerych - Mind to give it another look?

Changes:
- Local posts are now trashable and recoverable.
- Remote posts with local changes are trashable and recoverable.
- Whenever there are local changes, both the revision and the original are trashed / deleted / recovered together.

We still have some GUI refreshing issues, but I'd avoid treating those issues as part of this PR as those issues, even if annoying, are unrelated to the changes.

aerych commented:
Hiya @diegoreymendez 
I took the branch for another spin and noticed a few things. 

#### Missing prompt
Trashing a post while offline used to prompt that the internet connection was offline. Not seeing the prompt now.  I'm not sure if its related to changes in this PR but noting it here.

#### Local flag remains when published
The following affects drafts and scheduled posts. 
- Edit while offline so there s a local change. 
- On the post list, tap the more button to reveal the publish button. 
- Tap publish. 

The post is correctly published with the local changes, however the local flag remains. Maybe the revision isn’t removed when published?  When I open the editor the prompt for unsaved changes appears. 

#### Empty post when discarding changes.
There is some odd behavior discarding a local-only post. Try this:
- switch to airplane mode.
- create a new post while offline, and give it a title and some content. 
- notice it appears in the list as expected with the local status. 
- open the editor again and this time cancel the post and discard changes. 

Back in the list the post, the post is still there sans title or content.
<img width="378" alt="screen shot 2016-03-17 at 11 54 43 am" src="https://cloud.githubusercontent.com/assets/1435271/13854194/48709f50-ec38-11e5-85dc-4daef5237d80.png">
Maybe in the case of a new post started offline, changes in the revision should be saved to the original the very first time the post is saved. 


Interestingly, the issues in #4968 and #4969 do not occur when a draft or scheduled post has local changes.

diegoreymendez commented:
@aerych - Thanks for the feedback, and sorry for the continuous issues - this is proving more complex that I had initially thought.  I'll address all your points and reassign when ready.

diegoreymendez commented:
@aerych - I've just updated the PR.

> Missing prompt

I haven't purposely changed anything on regards to this, but I noticed the same as you.  I believe this is a regression in my PR.

That said, maybe it'd be the right time to evaluate if we want to keep the popup as it is.  I believe that if we want the app to be friendlier for offline use, we may want to avoid showing the popup (or at least avoid showing it more than once, until the connection comes back up).  The popup requires user interaction to go away and interrupts what you're doing, even when you know the app is offline.

What about just showing the error message in the black overlay and removing the popup?

> Local flag remains when published

I fixed this.  I'm also thinking `PostService` could use some maintenance after I'm done with this PR.

> Empty post when discarding changes.

Should be fixed too.

diegoreymendez commented:
/cc @SergioEstevao - would you mind joining in too for the reviews?

diegoreymendez commented:
/cc @rachelmcr - In case you can also test this (not sure if you can).

diegoreymendez commented:
Closing because there's a `release/6.1` branch now.

Follow up is: https://github.com/wordpress-mobile/WordPress-iOS/pull/5009
